File: src/md_cat.php
<?php

namespace WpRefs\MdCat;
/*

use function WpRefs\MdCat\add_Translated_from_MDWiki;
use function WpRefs\MdCat\get_url_curl;

*/

use function WpRefs\TestBot\echo_test;



function get_url_curl(string $url): string
{
    $usr_agent = 'WikiProjectMed Translation Dashboard/1.0 (https://mdwiki.toolforge.org/; tools.mdwiki@toolforge.org)';

    $ch = curl_init();

    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    // curl_setopt($ch, CURLOPT_COOKIEJAR, "cookie.txt");
    // curl_setopt($ch, CURLOPT_COOKIEFILE, "cookie.txt");

    curl_setopt($ch, CURLOPT_USERAGENT, $usr_agent);

    curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);
    curl_setopt($ch, CURLOPT_TIMEOUT, 5);

    $output = curl_exec($ch);
    if ($output === FALSE) {
        echo_test("<br>cURL Error: " . curl_error($ch) . "<br>$url");
    }

    curl_close($ch);

    return $output;
}

function load_from_local_file()
{
    $localFile = dirname(__DIR__) . '/resources/mdwiki_categories.json';
    $decoded = [];
    if (is_file($localFile)) {
        $decoded = json_decode(file_get_contents($localFile) ?: '[]', true);
    }
    return $decoded;
}
function get_cats()
{
    $url = "https://www.wikidata.org/w/rest.php/wikibase/v1/entities/items/Q107014860/sitelinks";
    static $json = null;

    if (is_array($json)) {
        return $json;
    }

    $data = get_url_curl($url);
    $decoded = json_decode($data, true);

    if (!is_array($decoded) || empty($decoded)) {
        $decoded = load_from_local_file();
    }

    $json = is_array($decoded) ? $decoded : [];

    return $json;
}

function Get_MdWiki_Category($lang)
{
    // ---
    // https://it.wikipedia.org/w/index.php?title=Categoria:Translated_from_MDWiki&action=edit&redlink=1
    $skip_langs = [
        "it"
    ];
    // ---
    if (in_array($lang, $skip_langs)) {
        return "";
    }
    // ---
    $cats = get_cats();
    // ---
    $cat = $cats[$lang . "wiki"]["title"] ?? "Category:Translated from MDWiki";
    // ---
    return $cat;
}

function add_Translated_from_MDWiki($text, $lang)
{
    // ---
    if (preg_match("/:\s*Translated[ _]from[ _]MDWiki\s*\]\]/iu", $text)) {
        return $text;
    };
    // ---
    $cat = Get_MdWiki_Category($lang);
    // ---
    if (!empty($cat) && strpos($text, $cat) === false) {
        $text .= "\n[[$cat]]\n";
    }
    // ---
    return $text;
}

File: src/md_cat.py
"""MDWiki category helpers translated from PHP."""

from __future__ import annotations

import json
import re
from functools import lru_cache
from urllib.error import URLError
from urllib.request import Request, urlopen

from .TestBot import echo_test


USER_AGENT = "WikiProjectMed Translation Dashboard/1.0 (https://mdwiki.toolforge.org/; tools.mdwiki@toolforge.org)"


def get_url_curl(url: str) -> str:
    request = Request(url, headers={"User-Agent": USER_AGENT})
    try:
        with urlopen(request, timeout=5) as response:  # nosec: B310
            return response.read().decode("utf-8")
    except URLError as exc:  # pragma: no cover - network issues
        echo_test(f"\ncURL Error: {exc}\n{url}")
        return ""


@lru_cache(maxsize=1)
def get_cats() -> dict:
    url = "https://www.wikidata.org/w/rest.php/wikibase/v1/entities/items/Q107014860/sitelinks"
    data = get_url_curl(url)
    try:
        return json.loads(data or "{}")
    except json.JSONDecodeError:
        return {}


def Get_MdWiki_Category(lang: str) -> str:
    if lang in {"it"}:
        return ""
    cats = get_cats()
    return cats.get(f"{lang}wiki", {}).get("title", "Category:Translated from MDWiki")


def add_Translated_from_MDWiki(text: str, lang: str) -> str:
    if not text:
        return text
    if re.search(r":\s*Translated[ _]from[ _]MDWiki\s*\]\]", text, flags=re.IGNORECASE):
        return text
    cat = Get_MdWiki_Category(lang)
    if cat and cat not in text:
        text += f"\n[[{cat}]]\n"
    return text


__all__ = ["add_Translated_from_MDWiki", "Get_MdWiki_Category", "get_url_curl"]
